### PR TITLE
Only update the icon of a tab it the icon actually _changed_

### LINKED ISSUE #1333: Icons flash for a frame when switching tabs
Every time we change the focus of a tab/pane, we attempt to update the icon and titlebar of the tab. Since we're only storing a _path_ to the icon in memory, every time you change focus, we need to reload the icon from disk to create the `Icon`. 

This is obviously bad. we should instead be caching the icon once, and just updating the `Icon` with that pre-loaded one. Or at the very least, we should cache what the path to the last icon we set was, and only reload the icon if the path is different.

### LINKED ISSUE #2329: Windows Terminal crashes when a profile's icon URL is unparseable/invalid
I got the latest from choco and set up an Ubuntu terminal entry. I found that if I put a garbage URL in the profiles.json icon setting entry it crashes Windows Terminal and WT won't start thereafter. 

That probably shouldn't happen, though technically low priority its probably making WT look fragile. 

To repeat this behavior simply remove "ms-appx:" from any icon path.

### LINKED ISSUE #2339: Feature Request Option to add profiles to the Windows Start Menu
# Description of the new feature/enhancement

# Proposed technical implementation details (optional)

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

Only update the icon of a tab it the icon actually _changed_. The way we have it, we'll try and update the icon everytime focus changes, causing the tab icon to blink for a second, even when switching between two panes with the same profile.

 
## References

While I was here, I also fixed #2329, I think accidentally. But that's cool!

## PR Checklist  
* [x] fixes #1333
* [x] Also tangentially fixes #2329
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed (please review #2294 so I can start checking this box)

### CODE REVIEW

File: src/cascadia/TerminalApp/App.cpp [hunk 1]
```diff
@@ -928,7 +930,10 @@ namespace winrt::TerminalApp::implementation
         // Set this profile's tab to the icon the user specified
         if (profile != nullptr && profile->HasIcon())
         {
-            tabViewItem.Icon(_GetIconFromProfile(*profile));
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/App.cpp]: I don't necessarily love how many different places we need to do this now!

File: src/cascadia/TerminalApp/Tab.h [hunk 2]
```diff
@@ -37,6 +39,7 @@ class Tab
 
 private:
     std::shared_ptr<Pane> _rootPane{ nullptr };
+    winrt::hstring _lastIconPath{};
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/Tab.h]: i'd like to move towards only using hstring at projection boundaries -- it's less allocations and more manageable to use wstrings. thoughts?

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.h]: I'm tempted to make Tab a winrt type to have an ObservableProperty IconPath, so it'll just update automagically. So this is like, half way there

Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/Tab.h]: I reckon once we do that we can actually set the _tab object_ to be the "Header" in the TabViewItem.

File: src/cascadia/TerminalApp/App.cpp [hunk 3]
```diff
@@ -1252,19 +1251,22 @@ namespace winrt::TerminalApp::implementation
     {
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/App.cpp]: do we still need two copies of GetIconFromProfile? This code is duplicated in Tab.cpp now.

File: src/cascadia/TerminalApp/Tab.h (hunk 2 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/Tab.h]: my hugest sadness with `winrt::hstring` is that we _must_ (are contractually obligated to) use the less efficient type every time we use it.

`winrt::param::hstring` automatically handles wstring_view and wstring and stuff -- it does some magic to save allocations and everything. _It is also in a private namespace._

`winrt::hstring` forces an allocation of a real HSTRING ☹️

File: src/cascadia/TerminalApp/Tab.cpp [hunk 4]
```diff
@@ -153,25 +154,7 @@ void Tab::UpdateIcon(const winrt::hstring iconPath)
...
-                // non-transparent pixels in the image.
-                iconSource.ShowAsMonochrome(false);
-                iconSource.UriSource(iconUri);
-                elem.IconSource(iconSource);
-            }
-            CATCH_LOG();
-        }
-
-        _tabViewItem.Icon(elem);
+        _tabViewItem.Icon(GetColoredIcon(_lastIconPath));
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/Tab.cpp]: This one lost its exception handler.. It also no longer checks whether the path was empty. Does this work right for transitioning from "an icon" to "no icon"?

**Comment by DHowett-MSFT** [comment]
How does this fix #2339?

**Comment by zadjii-msft** [comment]
@DHowett-MSFT that was a typo, definitely meant #2329

File: src/cascadia/TerminalApp/Tab.cpp (hunk 4 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/Tab.cpp]: The exception handler went up into GetColoredIcon, but I'll double check the icon->null path

**Comment by DHowett-MSFT** [comment]
In that case, we can remove the exception handler around the otehr call. Do we need `CATCH_LOG` (it'll generate a report in the debugger)? or do we want a different type of catch (quieter kind)?

**Comment by DHowett-MSFT** [comment]
It's baa~ack!

```
 fatal error C1083: Cannot open compiler intermediate file:
'd:\a\1\s\obj\x64\auditmode\localtests_terminalapp\terminalapp.localtests.pch': Not enough space
```

**Comment by AdamDanischewski** [comment]
#2329 is not fixed in v0.4.2382.0 

Try the repetition step listed in #2329. Looks like Windows Terminal could use a QA boost.

**Comment by Reelix** [comment]
Can confirm that #2329 is still broken in v0.5.2681.0

Simple Replication::
1.) Open profiles.json
2.) Change any profiles icon to
> "icon": "InvalidFileName.png",

3.) Save

**Pull request merged**